### PR TITLE
feat: add nucleotide-count with error checking

### PR DESCRIPTION
**Pull request by vaeng**
I added an idea to include error handling into our test-framework.

The error_data.csv includes malformed inserts, that should not go into the table and should be blocked by table constraints and checks.

If they are inserted anyway, the sqlite3 reports an error, which is caught by the bash file and reported in a .csv, which is in turn re-read into the the final test table.

What dou you think?

### CODE REVIEW

**Review by vaeng — COMMENTED** [review]
There is also json in this exercise

**Comment by glennj** [comment]
We didn't get a chance to discuss this?

File: docs/LEARNING.md [hunk 1]
```diff
@@ -1,13 +1,15 @@
 # Learning
 
-<!-- TODO: write document
```
Comment (IsaacG) [inline docs/LEARNING.md:3]: These files are set up by the other PR. This looks like a duplicate change.

File: exercises/practice/nucleotide-count/nucleotide-count.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
+-- Task:  - Update the nucleotide_count table and set the result based on the input field.
+--        - Update table creation with constraints.
+
+DROP TABLE IF EXISTS nucleotide_count;
+CREATE TABLE "nucleotide_count" (
+    "strand" TEXT,
+    "result" TEXT,
+);
+
+-- Please don't change those import lines:
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/nucleotide-count.sql:10]: Should we also ask they not change the drop/create? Is there a reason that's not done in create_fixture.sql?

File: exercises/practice/nucleotide-count/.meta/example.sql [hunk 3]
```diff
@@ -0,0 +1,31 @@
+-- Task:  - Update the nucleotide_count table and set the result based on the input field.
+--        - Update table creation with constraints.
+
+DROP TABLE IF EXISTS nucleotide_count;
+CREATE TABLE "nucleotide_count" (
+    "strand" TEXT,
+    "result" TEXT,
+    CHECK  (1 >  LENGTH(REPLACE(REPLACE(REPLACE(REPLACE(strand, 'A', ''), 'C', ''), 'G', ''), 'T', '')))
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/.meta/example.sql:8]: Line 11 suggests the default CREATE TABLE should not be modified.

File: exercises/practice/nucleotide-count/.meta/example.sql [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        LENGTH(strand) - LENGTH(REPLACE(strand, 'G', '')),
+        LENGTH(strand) - LENGTH(REPLACE(strand, 'T', ''))
+    FROM nucleotide_count;
+
+UPDATE nucleotide_count
+SET result = (
+    SELECT json(' { "A": ' || a || ', "C": ' || c || ', "G": ' || g || ', "T": ' || t || ' } ')
+    FROM counts
+    WHERE nucleotide_count.strand == counts.strand
+        );
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/.meta/example.sql:31]:
```suggestion
);
```

File: exercises/practice/nucleotide-count/error-checking [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+# The csv can than be imported into the tests table to be shown to students.
+
+# Arguments:
+# $1: exercise slug
+
+# Output:
+# Writes the test results to a check_errors.csv file.
+
+slug=$1
+snake_slug=$(echo "${slug}" | sed 's/-/_/g')
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/error-checking:14]:
```suggestion
snake_slug="${slug//-/_}"
```

File: exercises/practice/nucleotide-count/error-checking [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+# prepare input file:
+cat ./create_fixture.sql > temp_sql_input.sql
+printf "\n" >> temp_sql_input.sql
+cat "./${slug}.sql" >> temp_sql_input.sql
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/error-checking:19]:
```suggestion
{
    cat ./create_fixture.sql
    echo
    cat "./${slug}.sql"
} > temp_sql_input.sql
```

File: exercises/practice/nucleotide-count/error-checking [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+rm check_errors.csv -f
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/error-checking:21]:
```suggestion
[[ -f check_errors.csv ]] && rm check_errors.csv
```

File: exercises/practice/nucleotide-count/error-checking [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+do
+    status="pass"
+    message=""
+    task_id=0
+    test_code=""
+    output=""
+    test_code="INSERT INTO ${snake_slug} (${column}) VALUES ('${insert}');"
+    sqlite_output=$(sqlite3 '' --init temp_sql_input.sql -bail  "${test_code}" ".exit" 2>&1)
+    if [ $? -eq 0 ] 
+    then
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/error-checking:32]:
```suggestion
    if sqlite3 -bail "${test_code}" < temp_sql_input.sql  ".exit" >/dev/null 2>&1
    then
```

File: exercises/practice/nucleotide-count/error-checking [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        status="fail"
+        message="${insert} should not be inserted into the table"
+    fi
+    # import exptects 9 values:
+    echo "${uuid},${name},${status},${message},${output},${test_code},${task_id},," >> check_errors.csv
+done < error_data.csv
+
+# cleanup
+rm temp_sql_input.sql -f
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/error-checking:41]:
```suggestion
rm temp_sql_input.sql
```

File: exercises/practice/nucleotide-count/error-checking (hunk 5 above)
Comment (glennj) [reply exercises/practice/nucleotide-count/error-checking:14]: I've been wondering why we kebab-snake names: `nucleotide-count_test.sql`. Should we not stick with one style?

**Comment by glennj** [comment]
I dunno, I find this is really convoluted.  We have tests we expect to pass, and if they don't pass it's a fail, and we have tests we expect to fail and if they don't fail it's a fail.

It would seem more straightforward to expect a JSON object like `{"error": "invalid nucleotide"}`

Besides, the error-checking scripts looks like it contains a classic code injection attack.

File: exercises/practice/nucleotide-count/error-checking (hunk 6 above)
Comment (vaeng) [reply exercises/practice/nucleotide-count/error-checking:19]: Ah, nice.

File: exercises/practice/nucleotide-count/nucleotide-count.sql (hunk 2 above)
Comment (vaeng) [reply exercises/practice/nucleotide-count/nucleotide-count.sql:10]: With SQLite you cannot alter the table and add checks after it was created, so the checks have to come into place when the table is created.

Or you have to rename it, make a new table with the checks, and copy the data.

File: exercises/practice/nucleotide-count/error-checking (hunk 5 above)
Comment (vaeng) [reply exercises/practice/nucleotide-count/error-checking:14]: Yes, we should keep it one way.
We had one commit, where the case was changed, but only for the table name.
I could not find it, we should have them all the same.

Are there arguments for kebab or snake? I would prefer to have everything in snake case. But I have no real reason.

**Comment by vaeng** [comment]
> I dunno, I find this is really convoluted. We have tests we expect to pass, and if they don't pass it's a fail, and we have tests we expect to fail and if they don't fail it's a fail.
> 
> It would seem more straightforward to have all the inputs in data.csv and to expect a JSON object like `{"error": "invalid nucleotide"}` for bad inputs.
> 
> Besides, the error-checking scripts looks like it contains a classic code injection attack (calling Bobby Tables).

I am not super happy with the solution, but I thought it would be the most "SQL" idiomatic way to block invalid data from the table with restraints and checks.
I might change the whole bash thing to a trigger, that detects if the data was inserted or not? But I had no success in finding the best syntax. Upsert seemed like a cool candidate, but it does not seem to work on different tables (if it cannot insert into table a, please update table b).

**Comment by glennj** [comment]
> block invalid data from the table with restraints and checks.

We could provide all the cases in a `strands` table, and then it's up to the student to move to a "valid" table and an "invalid" table
```sql
create table strands (strand text);
.mode csv
.import ./all_data.csv strands

create table valid_strands (strand text, counts text);
create table invalid_strands (strand text);

update valid_strands v
set v.strand = raw_data.strand
  , v.counts = ...
from raw_data
where ...
;

update invalid_strands i
set i.strand = raw_data.strand
from raw_data
where NOT ...
;
```

**Comment by vaeng** [comment]
@glennj @IsaacG 

I redid the error checking mechanism. It is now all included in the _test.sql file and does not need any external additions.

The principle is as follows:
- problem spec cases that should produce an error have error as expected.
- this error translates to 'errror' in the expected field of those cases in our tests table.
- all those cases start with 'pass'
- their input is then inserted into the nucleotid_count table
- with the (student-created) checks in that table, insertions should be blocked
- a trigger (after insertion) is only fired if those entries made it into the table
- this trigger sets the respective test to 'fail'

File: exercises/practice/nucleotide-count/error-checking (hunk 5 above)
Comment (IsaacG) [reply exercises/practice/nucleotide-count/error-checking]: I'm leaning towards using kebab everywhere so the slug and filenames match up. If we use snake, we need to juggle both, one of the slug, the other for the file. `exercises/practice/${kebab}/${snake}.sql`.

File: exercises/practice/nucleotide-count/nucleotide-count.sql [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+CREATE TABLE "nucleotide_count" (
+    "strand" TEXT,
+    "result" TEXT
+);
+
+-- Please don't change the following two import lines:
+.mode csv
+.import ./data.csv nucleotide_count
+
+-- Write your code below:
```
Comment (IsaacG) [inline exercises/practice/nucleotide-count/nucleotide-count.sql:14]:
```suggestion
-- Write your code below. Feel free to edit the CREATE TABLE above, too!
```

**Pull request merged**